plugins/Jobs.cpp

```diff
@@ -550,35 +592,35 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
         const string& name = request["name"];
         string safeNumResults = SQ(max(request.calc("numResults"),1));
         string selectQuery =
-            "SELECT jobID, name, data, parentJobID FROM ( "
+            "SELECT jobID, name, data, parentJobID, retryAfter FROM ( "
                 "SELECT * FROM ("
-                    "SELECT jobID, name, data, priority, parentJobID "
+                    "SELECT jobID, name, data, priority, parentJobID, retryAfter "
                     "FROM jobs "
-                    "WHERE state='QUEUED' "
+                    "WHERE state IN ('QUEUED', 'RUNQUEUED') "
                     "  AND priority=1000"
                     "  AND " + SCURRENT_TIMESTAMP() + ">=nextRun "
                     "  AND name GLOB " + SQ(name) + " "
-                    "ORDER BY nextRun ASC LIMIT " + safeNumResults +
+                    "ORDER BY nextRun, jobID ASC LIMIT " + safeNumResults +
```

Comment:
1.  I think what you're asking here is if `ORDER BY priority DESC` does a stable sort.  Is that correct?  If so, I created tests for `GetJob` that test the return order of jobs with the same `priority` and `nextRun` but different  `state` values.  Therefore I'm confident that `order by nextRun, jobID` isn't affected by `ORDER BY priority DESC`.  Check out the `testPrioritiesWithRunQueued` test
1.  I haven't benchmarked this but I did check that the query plan is the same.  If you want, I can benchmark it tomorrow
Query plan without `jobID ASC`
```
~$ sudo readdb.sh "explain query plan SELECT jobID, name, data, priority, parentJobID, retryAfter
> FROM jobs
> WHERE state IN ('QUEUED', 'RUNQUEUED')
>   AND priority=0
>   AND current_timestamp >=nextRun
>   AND name GLOB 'www-prod/*'
> ORDER BY nextRun ASC LIMIT 10;"
selectid    order       from        detail
----------  ----------  ----------  -------------------------------------------------------------------------------------------------
0           0           0           SEARCH TABLE jobs USING INDEX jobsStatePriorityNextRunName (state=? AND priority=? AND nextRun<?)
0           0           0           EXECUTE LIST SUBQUERY 1
0           0           0           USE TEMP B-TREE FOR ORDER BY
```
Query plan with `jobID ASC`
```
~$ sudo readdb.sh "explain query plan SELECT jobID, name, data, priority, parentJobID, retryAfter
> FROM jobs
> WHERE state IN ('QUEUED', 'RUNQUEUED')
>   AND priority=0
>   AND current_timestamp >=nextRun
>   AND name GLOB 'www-prod/*'
> ORDER BY nextRun ASC, jobID ASC LIMIT 10;"
selectid    order       from        detail
----------  ----------  ----------  -------------------------------------------------------------------------------------------------
0           0           0           SEARCH TABLE jobs USING INDEX jobsStatePriorityNextRunName (state=? AND priority=? AND nextRun<?)
0           0           0           EXECUTE LIST SUBQUERY 1
0           0           0           USE TEMP B-TREE FOR ORDER BY
```

Reply:
> created tests for GetJob that test the return order of jobs with the same priority and nextRun but different state values. Therefore I'm confident that order by nextRun, jobID isn't affected by ORDER BY priority DESC

The test should be with the same priority and nextRun but different priorities not different states, right?

What I'm thinking with this is that we _may_ be slowing down one of the main queries to solve what seems to me like a non-isssue which is the fact that Job B which was created after Job A may be processed before Job A if they both have the same priority and the very same nextRun. The reason I call it a non-issue is because I understand the case but it definitely seems like an edge case and I haven't seen that this is causing a problem, am I missing something? Would you agree that this change might not be worth implementing now?

Reply:
If we generally consider this a non-issue, then I can revert this change.  The reason I implemented it was because this test failed
https://github.com/Expensify/Bedrock/pull/243/files#diff-c8ebb7cec8b361355acfa8a63051dacbR364

Now the following jobs also tests will also fail 
https://github.com/Expensify/Bedrock/pull/243/files#diff-c8ebb7cec8b361355acfa8a63051dacbR268
https://github.com/Expensify/Bedrock/pull/243/files#diff-cf845d40b06e01293a549aa13ac565adR326

So if this is a non-issue, I'll update the tests accordingly

Reply:
Actually, @quinthar do you have an opinion here? 
I agree with @cead22 that this is an edge case but it's also the proper way to achieve the behavior we want.
If we move forward with this change, I should devise a plan to test this query

Reply:
I'm not sure I follow.  Are you asking if it's worth processing same-priority jobs in order, or if it's ok if same-priority jobs execute out of order?  I'd say it would be great to process them in order if possible, and it sounds like @mea36's test confirms that it is not just possible, but has no effect on the query plan (and thus should have the same performance).  Am I misunderstanding something?

Reply:
No that's exactly right and for added context:
- I agree that Mel's implementation is correct
- I don't think that 2 equal query plans translates into two equal query execution times. My worry here specifically is that the temp btree used to sort could be bigger since it's now using 2 values and thus take longer to create and traverse (which we do 3 times)
- In light of that and our recent query perf degradation problems, I'd love for us to bechmark the new query vs the old one before releasing to be on the safer side
- And because of that I proposed that as an alternative we don't solve this edge case now given that this doesn't solve any problems from the user's perspective, and live with the fact that 2 jobs with the exact same nextRun and the exact same priority maybe be dispatched our of order.

Reply:
Since the hard work was already done to do some performance testing (thanks @cead22 ), I followed the steps in https://github.com/Expensify/Expensify/issues/49378#issuecomment-298134063 using a jobs table from @cead22 that has 347K jobs;

I created a new database and created a table using the schema I have on dev (minus the `retryAfter` column
```sql
sqlite> .schema jobs
CREATE TABLE jobs ( created TIMESTAMP NOT NULL, jobID INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, state TEXT NOT NULL, name TEXT NOT NULL, nextRun TIMESTAMP NOT NULL, lastRun TIMESTAMP, repeat TEXT NOT NULL, data TEXT NOT NULL, priority INTEGER NOT NULL DEFAULT 500, parentJobID INTEGER NOT NULL DEFAULT 0);
CREATE INDEX jobsState    ON jobs ( state    );
CREATE INDEX jobsName     ON jobs ( name     );
CREATE INDEX jobsNextRun  ON jobs ( nextRun  );
CREATE INDEX jobsLastRun  ON jobs ( lastRun  );
CREATE INDEX jobsPriority ON jobs ( priority );
CREATE INDEX jobsParentJobIDState ON jobs ( parentJobID, state );
CREATE INDEX jobsStatePriorityNextRunName ON jobs ( state, priority, nextRun, name );
```

Then copied the subquery to a script that ran the query 1000 times

I ran this twice for each version of the query.
`ORDER BY nextRun ASC` had an average runtime of `0.003772` and `0.003973`
`ORDER BY nextRun ASC, jobID ASC` had an average runtime of `0.004018` and `0.003936`

I think the difference is inconsequential and I think we should move forward with these changes

Reply:
👍

Reply:
Looks like our performance testing wasn't adequate here -- when harvesting started, 1.5M jobs became runnable, and performance slowed to a near-halt.  What's interesting is we had more or less this exact fear, and actually thought we tested to confirm it, but still missed it.  What can we do differently next time to catch this?  Only do performance testing on prod, and only right after harvesting starts?

Reply:
Cc: @coleaeason

Reply:
Well one way to catch this next time would be to make sure we time queries while harvesting is running and while it's not running. 

Another way to catch this would be if we could build a test that mimics harvesting, i.e. it queues 1.5mn run ready jobs that have similar parameters to harvesting jobs (retry times, data, etc.). It would be nice if we had a way to fail a test if a query or operation takes more time than X.

Reply:
> Well one way to catch this next time would be to make sure we time queries while harvesting is running and while it's not running.

I agree, running queries in the jobs table when the queue is almost empty is almost as running them on dev. 
Not sure how big is the jobs table, but maybe we can grab a snapshot of it when it has data and save it to use it in any performance test we need.